dao/application_dao.go

```diff
@@ -170,6 +172,15 @@ func (a *applicationDaoImpl) Create(app *m.Application) error {
 	app.TenantID = *a.TenantID
 	result := DB.Debug().Create(app)
 
+	// Check if specific error code is returned
+	switch e := result.Error.(type) {
+	case *pgconn.PgError:
+		// unique constraint violation for index (source id + app type id + tenant id)
+		if e.Code == uniqueConstraintViolation && strings.Contains(e.Detail, "Key (source_id, application_type_id, tenant_id)") {
+			message := fmt.Sprintf("Application of application type = %d already exists for the source id = %d", app.ApplicationTypeID, app.SourceID)
+			return util.NewErrBadRequest(message)
+		}
+	}
```

Comment:
So rather than switching on the type - we could just use `errors.As` which trys to cast the error for us and returns true if and only if the error is the right type, this is what it would look like:
```suggestion
	var pgErr *pgconn.PgError
	if errors.As(result.Error, &pgErr) {
		// unique constraint violation for index (source id + app type id + tenant id)
		if pgErr.Code == uniqueConstraintViolation && strings.Contains(pgErr.Detail, "Key (source_id, application_type_id, tenant_id)") {
			message := fmt.Sprintf("Application of application type = %d already exists for the source id = %d", app.ApplicationTypeID, app.SourceID)
			return util.NewErrBadRequest(message)
		}
	}
```

Found here: https://github.com/jackc/pgx/wiki/Error-Handling
I wish there were consts for the error codes in the package but unfortunately it doesn't look like the case since this is a pretty low-level database adapter.

Reply:
thanks 😁 done